toolkit/smart-contracts/offchain/src/multisig.rs

```diff
@@ -0,0 +1,160 @@
+	pub temporary_wallet: TemporaryWalletData,
+	#[serde(serialize_with = "serialize_as_conway_tx")]
+	pub tx_cbor: Vec<u8>,
+}
+
+/// To be used only for manual re-claim of the funds if transaction has not been submitted
+#[derive(Clone, Debug, Serialize)]
+pub struct TemporaryWalletData {
+	pub address: String,
+	pub private_key: ByteString,
+	pub funded_by_tx: McTxHash,
+}
+
+pub(crate) struct TemporaryWallet {
+	pub address: cardano_serialization_lib::Address,
+	pub private_key: CardanoPaymentSigningKey,
+	pub funded_by_tx: [u8; 32],
+}
+
+impl From<TemporaryWallet> for TemporaryWalletData {
+	fn from(value: TemporaryWallet) -> Self {
+		TemporaryWalletData {
+			address: value.address.to_bech32(None).unwrap(),
+			private_key: value.private_key.to_bytes().into(),
+			funded_by_tx: McTxHash(value.funded_by_tx),
+		}
+	}
+}
+
+fn serialize_as_conway_tx<S>(tx_bytes: &Vec<u8>, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	let json = serde_json::json!({
+		"type": "Tx ConwayEra",
+		"description": "",
+		"cborHex": hex::encode(tx_bytes)
+	});
+	json.serialize(serializer)
+}
+
+impl MultiSigSmartContractResult {
+	pub fn tx_submitted(hash: [u8; 32]) -> Self {
+		Self::TransactionSubmitted(McTxHash(hash))
+	}
+}
+
+pub(crate) async fn create_temporary_wallet<
+	T: QueryLedgerState + Transactions + QueryNetwork + QueryUtxoByUtxoId,
+	A: AwaitTx,
+>(
+	ctx: &TransactionContext,
+	client: &T,
+	await_tx: &A,
+) -> anyhow::Result<TemporaryWallet> {
+	let private_key = CardanoPaymentSigningKey(PrivateKey::generate_ed25519()?);
+	log::info!(
+		"Temporary wallet private key: {}. Store this key for eventual re-claim of tokens",
+		private_key.0.to_hex()
+	);
+	let address = key_hash_address(&private_key.0.to_public().hash(), ctx.network);
+
+	let mut funding_tx_builder = TransactionBuilder::new(&get_builder_config(ctx)?);
+	// TODO: ETCM-9627 - estimate tokens needed for the transaction
+	funding_tx_builder
+		.add_output(&TransactionOutput::new(&address, &Value::new(&5_000_000u32.into())))?;
+	let funding_tx = funding_tx_builder.balance_update_and_build(ctx)?;
+	let funding_tx_result = client.submit_transaction(&ctx.sign(&funding_tx).to_bytes()).await?;
+	let funded_by_tx = funding_tx_result.transaction.id;
+	await_tx.await_tx_output(client, UtxoId::new(funded_by_tx, 0)).await?;
+	let address_str: String = address.to_bech32(None)?;
+	log::info!(
+		"Founded temporary wallet {} with 5 ADA in transaction: {}",
+		&address_str,
+		&hex::encode(funded_by_tx)
+	);
+	Ok(TemporaryWallet { address, private_key, funded_by_tx })
+}
+
+/// If the chain has real MultiSig governance it:
+/// * creates a temporary wallet
+/// * sends 5 ADA from the payment wallet (subject of change)
+/// * creates a transaction that would be paid from the temporary wallet, signed by both wallets.
+/// If the chain has single key governance it creates and submits transaction paid by and signed by the payment wallet.
+pub(crate) async fn multisig_process<F, T, A>(
```

Comment:
This function processes either multisig or single keys but is called `multisig_process` - that's maybe not ideal? That being said only calling it `process` is probably too generic.

Reply:
From day 1 of offchain we supported only Multisig Policy. But instead of treating multisig with 1 of 1 as special case, it was our default and only case. Recent changes are using native Cardano MultiSig and adding more keys.

The function itself won't work if the on-chain governance is not native MultiSig or Plutus MultiSig policy from partner-chains-smart-contracts repo. I'll add this remark to docs. I'll give myself few minutes to rethink the name of this function.

Reply:
Correction: this function cannot be even invoked with not multisig policy, because Governance Data policy is either `PartnerChainsMultisigPolicy` or `SimpleAtLeastN`